spec/lib/api/2.0/templates-spec.js

```diff
@@ -0,0 +1,225 @@
+            return helper.request().get('/api/2.0/templates/metadata/123')
+            .expect('Content-Type', /^application\/json/)
+            .expect(200, template)
+            .then(function() {
+                expect(templates.getName).to.have.been.calledWith('123');
+            });
+        });
+
+        it('should return 404 for invalid templates name', function() {
+            templates.get.rejects(new Errors.NotFoundError('bad_template'));
+            return helper.request().get('/api/2.0/templates/library/test')
+            .expect(404)
+            .then(function() {
+                expect(templates.get).to.have.been.calledOnce;
+                expect(templates.get).to.have.been.calledWith('test');
+            });
+        });
+
+    });
+
+    describe('2.0 PUT /templates/library/:name', function () {
+        beforeEach('should PUT new mockfile', function () {
+            return helper.request().put('/api/2.0/templates/library/testTemplate')
+                .send('test\n')
+                .expect(200)
+                .expect(function(){
+                    expect(templates.put).to.have.been.calledOnce;
+                    expect(templates.put).to.have.been.calledWith('testTemplate');
+                });
+        });
+
+        it('should get new templates', function () {
+            return helper.request().get('/api/2.0/templates/metadata/testTemplate')
+            .expect(200)
+            .then(function () {
+                expect(templates.getName).to.have.been.calledOnce;
+                expect(templates.getName).to.have.been.calledWith('testTemplate');
+            });
+        });
+
+        it('should 400 error when templates.put() fails', function () {
+            templates.put.rejects(new Error('dummy'));
+            return helper.request().put('/api/2.0/templates/library/123')
+            .send('test_template_foo\n')
+            .expect('Content-Type', /^application\/json/)
+            .expect(400);
+        });
+    });
+
+    describe('GET SB /templates/:name', function () {
+        it('should return a template', function () {
+            var graph = {
+                instanceId: '0123'
+            };
+            waterline.nodes = {
+                needByIdentifier: sinon.stub().resolves({
+                    id: 'node id'
+                })
+            };
+            findActiveGraphForTarget.resolves(graph);
+            this.sandbox.stub(swagger, 'makeRenderableOptions').resolves({});
+            taskProtocol.requestProperties.resolves({});
+            return helper.request().get('/api/2.0/templates/123')
+            .expect(200)
+            .then(function () {
+                expect(templates.render).to.have.been.calledOnce;
+                expect(templates.render).to.have.been.calledWith('123');
+            });
+        });
+    });
+
+    describe('2.0 DELETE /templates/library/:name', function () {
+        beforeEach('should PUT new mockfile', function () {
```

Comment:
I remembered mocha doesn't promise the execution sequence of all test cases within a test suite, mocha can sort out test cases, filtering test cases, so you cannot expect mocha will always run in your designed order: PUT --> GET -> DELETE -> GET ERROR.

So the each test case should be independent, all sequence should use the hook function `before`/`after`/`beforeEach`/`afterEach`.